CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h

```diff
@@ -0,0 +1,32 @@
+#ifndef CUDADataFormats_SiPixelCluster_interface_gpuClusteringConstants_h
+#define CUDADataFormats_SiPixelCluster_interface_gpuClusteringConstants_h
+
+#include <cstdint>
+
+namespace pixelGPUConstants {
+#ifdef GPU_SMALL_EVENTS
+  constexpr uint32_t maxNumberOfHits = 24 * 1024;
+#else
+  constexpr uint32_t maxNumberOfHits =
+      48 * 1024;  // data at pileup 50 has 18300 +/- 3500 hits; 40000 is around 6 sigma away
+#endif
+}  // namespace pixelGPUConstants
+
+namespace gpuClustering {
```

Comment:
@slava77 

> please clarify here or later the expected phase space of pixel local reco.

I think that the Patatrack team has tested this in simulation with the PU profile of Run3 we currently have in release (flat 55-75): https://github.com/cms-sw/cmssw/commit/f735efeca5c0fe6d13f279cd3b8d315242785a2c
this would be OK in terms of Run3. @AdrianoDee @VinInn please confirm.

For HI as well as for Phase-2 this is indeed not OK, but as we discussed privately via e-mail.
For HI we're aware of this discussion: https://github.com/cms-patatrack/cmssw/pull/545  please clarify if this is needed now or can be addressed later, after this is merged.
For Phase-2 anyway there are several other shortcoming, first and foremost we lack a raw2digi step for lack of the final hardware specs....

Reply:
Indeed all the latest validations , since  https://github.com/cms-patatrack/cmssw/pull/449 , are run on Run3-like PU 55-75 flat samples and the buffers have been big enough.

Reply:
> Indeed all the latest validations , since [cms-patatrack#449](https://github.com/cms-patatrack/cmssw/pull/449) , are run on Run3-like PU 55-75 flat samples and the buffers have been big enough.

it would be nice then to update the comments in the code to be clear about this (wherever PU 50 is mentioned)

Reply:
>  For HI we're aware of this discussion: [cms-patatrack#545](https://github.com/cms-patatrack/cmssw/pull/545) please clarify if this is needed now or can be addressed later, after this is merged.

IMHO, the compiler macro variants are inconsequential because they require a separate release build. So, in a big picture, the ability to use this code outside of Run3 scope is quite unclear.

This cms-patatrack#545 has also changes to the types without macros, which perhaps has some impact on the data sizes and could affect performance. If they have to be done, it may be useful to update these uint16 -> uint32 even in this PR.

Reply:
Not "perhaps", see https://github.com/cms-patatrack/cmssw/pull/474

Reply:
The index type has been moved to 32 bits in https://github.com/cms-patatrack/cmssw/pull/583, with negligible impact on the performance.

Making all the sizes configurable at runtime should be possible, and I agree that it should be the goal to support Phase 2 as well as Phase 1 with minimal code changes.
However, it will require significant code changes in multiple packages, and it would benefit from the development of a more user friendly, constant-size (instead of static-size) SoA interface.

I would propose to create an issue to keep track of this and other future developments, and process with the integration as is.

Reply:
@slava77 if I understand correctly you were suggesting to move the definitions inside `namespace gpuClustering` to `namespace pixelGPUConstants`.

I am in favour of using less namespaces, but this may have a complicated interaction with the use of the `gpuClustering` namespace in [RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h](https://github.com/cms-patatrack/cmssw/tree/master/RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h) and [RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h](https://github.com/cms-patatrack/cmssw/tree/master/RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h).

If there is a suggestion for a common namespace that applies there as well, I can make the change.

Reply:
I think that a compile time maxNumberOfHits  (and even a maxNumberOfDigis) is better to be kept around (it exists for Legacy as well even if configurable) even if the number of hits will be determined at runtime, copied on CPU, and used everywhere needed. We could issue a warning as in Legacy and the event will be flagged.